File: backend/src/main/java/com/example/demo/repository/PostRepositoryImpl.java
package com.example.demo.repository;

import com.querydsl.jpa.impl.JPAQueryFactory;

import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Repository;

import com.example.demo.entity.User;
import com.example.demo.entity.QUser;

@Repository
@Transactional(readOnly = true)
public class PostRepositoryImpl implements PostRepositoryCustom {

    private JPAQueryFactory queryFactory;

    public PostRepositoryImpl(JPAQueryFactory queryFactory) {
        this.queryFactory = queryFactory;
    }

    @Override
    @Cacheable(value = "searchConditionCache", key = "'defaultCondition'")
    public List<User> searchByCondition() {
        return queryFactory.selectFrom(QUser.user)
                .where(QUser.user.password.contains("12345"))
                .fetch();

    }

}

File: backend/src/main/kotlin/com/example/demo/repository/PostRepositoryImpl.kt
package com.example.demo.repository

import com.example.demo.entity.QUser
import com.example.demo.entity.User
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Repository
import org.springframework.transaction.annotation.Transactional

@Repository
@Transactional(readOnly = true)
class PostRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : PostRepositoryCustom {

    @Cacheable(value = ["searchConditionCache"], key = "'defaultCondition'")
    override fun searchByCondition(): List<User> {
        return queryFactory.selectFrom(QUser.user)
            .where(QUser.user.password.contains("12345"))
            .fetch()
    }
}
